OpenMage/magento-lts
# Enhanced Google Analytics 4 Integration and Ecommerce Tracking Improvements

### Description (*)
Improve Google Analytics 4 integration and ecommerce tracking 
PS: Apologies for not splitting into single issues/PRs for better comparison/readability. I've had to refactor multiple parts and am short on time.

### New Features
- add debug mode to view real time events in analytics (only for developer ip)
- add user_id tracking

### Fixed Issues 
- follow guidelines for item_* props names from [GA4 Events Reference](https://developers.google.com/analytics/devguides/collection/ga4/reference/events?hl=it&client_type=gtag)
- remove unnecessary generic item_list_name
- fix begin_checkout events firing twice on order success page
- add index position in view_item_list
- added product category and removed some unecessary generics
- fix missing price value for some events
- fix missing sku value if is not set as "used_in_product_listing"
- wrap tracking in its own function like Universal Analytics
- add support for firecheckout to event purchase

### Manual testing scenarios (*)
1. Set Analytics 4 mode and measurement ID
2. Enable debug mode and view real-time events in GA4; events are also written in googleanalytics4.log (add your IP as a developer)
3. Enable user_id tracking, log in to frontend, and verify that it is set correctly with customer_id

### Questions or comments
- Is it safe and in compliance with GA4 guidelines to send a customer ID as user_id?
- If the manufacturer attribute is not set to "used_in_product_listing", the value is empty. We haven't fixed this for performance reasons, but maybe we can add some info like "To track item_brand, set the manufacturer to 'used_in_product_listing'"?
- For the manufacturer, we can add a configuration to select another attribute
- For performance reasons, we've added the category only if it's "ready" without other loaders

We have tested it at some production sites, and it's working correctly for us as is.
